oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/AsyncCheckpointCreator.java

```diff
@@ -0,0 +1,121 @@
+        this.checkpointLifetimeInSeconds = checkpointLifetimeInSeconds;
+        this.minConcurrentCheckpoints = minConcurrentCheckpoints;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    protected long getCheckpointLifetimeInSeconds() {
+        return checkpointLifetimeInSeconds;
+    }
+
+    protected long getMinConcurrentCheckpoints() {
+        return minConcurrentCheckpoints;
+    }
+
+    @Override
+    public void run() {
+        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        String creationTimeStamp = String.valueOf(cal.getTimeInMillis());
+        String creationTimeISOFormat = ISO8601.format(cal);
+        String checkpoint = store.checkpoint(checkpointLifetimeInSeconds * 1000, Map.of(
+                CHECKPOINT_CREATOR_KEY, AsyncCheckpointCreator.class.getSimpleName(),
+                CHECKPOINT_CREATED_KEY, creationTimeISOFormat,
+                CHECKPOINT_CREATED_TIMESTAMP_KEY, creationTimeStamp,
+                CHECKPOINT_THREAD_KEY, Thread.currentThread().getName(),
+                CHECKPOINT_NAME_KEY, name));
+        log.info("[{}] Created checkpoint {} with creation time {}", name, checkpoint, creationTimeISOFormat);
+
+        // Get a list of checkpoints filtered on the basis of CHECKPOINT_NAME_KEY (name). This is done using the
+        // getFilteredCheckpoints in the IndexUtils, which gets all the checkpoints in the node store and then filters the list based on
+        // the provided predicate using the checkpoint info map associated with the checkpoints.
+        // The filtered checkpoints list is then sorted based on the CHECKPOINT_CREATED_TIMESTAMP_KEY (created-epoch).
+        // Both the initial filtering and sorting is done based on the information from the associated checkpoint info map of a given checkpoint.
+        Set<String> sortedCheckpointSet = IndexUtils.getSortedCheckpointMap(IndexUtils.getFilteredCheckpoints(store,
+                entry -> name.equals(entry.getValue().get(CHECKPOINT_NAME_KEY))), CHECKPOINT_CREATED_TIMESTAMP_KEY).keySet();
+        int counter = sortedCheckpointSet.size();
+        // Iterate over the sortedCheckpointSet which is sorted based on the creation timestamp -> the oldest first
+        // We try and delete the checkpoints as long as the counter is greater than minConcurrentCheckpoints
+        // This ensures that the system always has concurrent checkpoints equal to or greater than the configured minConcurrentCheckpoints
+        for (String cp : sortedCheckpointSet) {
+            // Delete the checkpoint as long as the checkpoint count is greater than concurrentCheckpoints
+            if (counter > minConcurrentCheckpoints) {
+                if(store.release(cp) ) {
+                    log.info("[{}] Deleted checkpoint {}", name, cp);
+                } else {
+                    log.warn("[{}] Unable to delete checkpoint {}. Removal will be attempted again in next run.", name, cp);
```

Comment:
If we are not able to release a checkpoint we retry in the next run. That's fine.
What if this condition persists for several runs? In this case, we will keep adding checkpoints without deleting them.
Should we add some logic before creating a new checkpoint which checks there are not too many checkpoints?

Reply:
@fabriziofortino  - the point to deliberate here is - what is worse - 

1. Having more number of orphaned checkpoints in the system
2. Having to do a larger checkpoint diff (because we didn't create newer checkpoints because older ones didn't get deleted).

Please note that the checkpoint lifetime is configurable, with default suggested value as 24 hours, so if a checkpoint somehow manages to not get deleted in consequent runs, it should get cleaned up after 24 hours itself (if we keep that value in the configurations).

Reply:
I can still possibly add a check to not create any checkpoints if the number of checkpoints in the system right now are let's say 10 times than that of the minConcurrentCheckpoints value ?

Reply:
> Or we could make it configurable with a value of maxConcurentCheckpoints as well
+1

Reply:
done